### PR TITLE
PT 1.12 Patching

### PR DESCRIPTION
**Pull request by kace**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 1]
```diff
@@ -316,7 +316,7 @@ def test_pytorch_training_torchdata_gpu(
     _, image_framework_version = get_framework_and_version_from_tag(pytorch_training)
     if test_utils.is_image_incompatible_with_instance_type(pytorch_training, ec2_instance_type):
         pytest.skip(f"Image {pytorch_training} is incompatible with instance type {ec2_instance_type}")
-    if Version(image_framework_version) in SpecifierSet("==1.11.*"):
+    if Version(image_framework_version) in SpecifierSet("==1.12.*"):
```
Comment (saimidu) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:319]: This seems to have previously been `PT_TORCHDATA_DEV_CMD` for PT 1.11, and `PT_TORCHDATA_CMD` for all others, which was introduced when adding PT 1.12 DLC images.

Why is this being changed to PT 1.12 now?

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 2]
```diff
@@ -333,7 +333,7 @@ def test_pytorch_training_torchdata_cpu(
...
```
Comment (saimidu) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:336]: This seems to have previously been `PT_TORCHDATA_DEV_CMD` for PT 1.11, and `PT_TORCHDATA_CMD` for all others, which was introduced when adding PT 1.12 DLC images.

Why is this being changed to PT 1.12 now?

Comment (kace) [reply test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:336]: For testing as mentioned above. I'll make another commit that supports both PT 1.11 and PT 1.12.

**Comment by kace** [comment]
Tests passing in [6572837](https://github.com/aws/deep-learning-containers/pull/2406/commits/65728376f4591b6efd62a2ea2d1e5558181d3509).

**Review by ohadkatz — APPROVED** [review]

File: test/dlc_tests/container_tests/bin/pytorch_tests/testTorchdata [hunk 3]
```diff
@@ -1,7 +1,5 @@
 #!/bin/bash
 
-set -e
-
```
Comment (saimidu) [inline test/dlc_tests/container_tests/bin/pytorch_tests/testTorchdata:4]: Why merge this in with `set -e` removed?

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 4]
```diff
@@ -316,7 +316,7 @@ def test_pytorch_training_torchdata_gpu(
     _, image_framework_version = get_framework_and_version_from_tag(pytorch_training)
     if test_utils.is_image_incompatible_with_instance_type(pytorch_training, ec2_instance_type):
         pytest.skip(f"Image {pytorch_training} is incompatible with instance type {ec2_instance_type}")
-    if Version(image_framework_version) in SpecifierSet("==1.11.*"):
+    if any(Version(image_framework_version) in v for v in [SpecifierSet("==1.11.*"), SpecifierSet("==1.12.*")]):
```
Comment (saimidu) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:319]: You can actually use:
```suggestion
if Version(image_framework_version) in SpecifierSet(">=1.11,<1.13")
```
instead.

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 5]
```diff
@@ -333,7 +333,7 @@ def test_pytorch_training_torchdata_cpu(
...
```
Comment (saimidu) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:336]: You can actually use:
```suggestion
if Version(image_framework_version) in SpecifierSet(">=1.11,<1.13")
```
instead.

File: test/dlc_tests/container_tests/bin/pytorch_tests/testTorchdataDev [hunk 6]
```diff
@@ -1,17 +1,16 @@
...
-set -e
-
 HOME_DIR=/test
 BIN_DIR=${HOME_DIR}/bin
 LOG_DIR=${HOME_DIR}/logs
 
-TORCHDATA_DEV_BRANCH="pt1.11.0-dlc-torchdata"
-TORCHDATA_DEV_REPO="https://github.com/ydaiming/data.git"
+TORCHDATA_DEV_BRANCH_PREFIX="torchdata_pt_"
+TORCHDATA_DEV_REPO="https://github.com/kace/data.git"
```
Comment (saimidu) [inline test/dlc_tests/container_tests/bin/pytorch_tests/testTorchdataDev:8]: Thanks!

Comment (kace) [reply test/dlc_tests/container_tests/bin/pytorch_tests/testTorchdataDev:8]: 😄

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py (hunk 5 above)
Comment (kace) [reply test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:336]: Good point. As this is already merged, I'll submit another PR to clean this up.

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py (hunk 4 above)
Comment (kace) [reply test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:319]: Ack'ed below.

File: test/dlc_tests/container_tests/bin/pytorch_tests/testTorchdata (hunk 3 above)
Comment (kace) [reply test/dlc_tests/container_tests/bin/pytorch_tests/testTorchdata:4]: The test author's original intention was for the script to execute from start to end since there is logging code after the test run.
